### PR TITLE
refactor(utils): reduce scoring and sorting type, fix logic

### PR DESCRIPTION
**Pull request by Tlacenka**
In this PR I fix and clean up scoring and sorting logic:
- Types and constants are reduced and moved to a separate file.
- Group and category reference sorting logic has been fixed.
- Unnecessary properties are removed from reports.
- Unit tests are added for the new/modified functions.

Note: It is getting a bit tedious to maintain the benchmarks for scoring as they use the same types.

### CODE REVIEW

**Review by BioPhoton — APPROVED** [review]
Nice clean up 🧼

**Review by BioPhoton — APPROVED** [review]
Lgtm

File: packages/utils/perf/score-report/optimized0.ts [hunk 1]
```diff
@@ -17,16 +18,11 @@ function groupRefToScore(audits: EnrichedAuditReport[]) {
...
-  groups: EnrichedScoredGroup[],
-) {
+function categoryRefToScore(audits: AuditReport[], groups: ScoredGroup[]) {
   return (ref: CategoryRef) => {
     switch (ref.type) {
       case 'audit':
-        const audit = audits.find(
-          a => a.slug === ref.slug && a.plugin === ref.plugin,
-        );
+        const audit = audits.find(a => a.slug === ref.slug);
```
Comment (matejchalk) [inline packages/utils/perf/score-report/optimized0.ts:25]: Why was this changed? The audit `slug` is not unique, you have to check the `plugin` slug as well.

(Same goes for `case 'group'`.)

File: packages/utils/src/lib/reports/types.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+export type SortableGroup = ScoredGroup & {
+  weight: number;
+  plugin: string;
+};
+
+export type SortableAuditReport = AuditReport & {
+  weight: number;
+  plugin: string;
+};
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/types.ts:30]: Good naming :+1:

File: packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap [hunk 3]
```diff
@@ -15,14 +15,18 @@ exports[`generateMdReport > should contain all sections when using the fixture r
...
 - 🟢 Performance (_Lighthouse_)
   - 🟩 [Total Blocking Time](#total-blocking-time-lighthouse) - **0 ms**
   - 🟨 [Largest Contentful Paint](#largest-contentful-paint-lighthouse) - **1.5 s**
   - 🟩 [Cumulative Layout Shift](#cumulative-layout-shift-lighthouse) - **0**
   - 🟨 [First Contentful Paint](#first-contentful-paint-lighthouse) - **1.2 s**
   - 🟩 [Speed Index](#speed-index-lighthouse) - **1.2 s**
 
+- 🔴 ESLint performance rules (_ESLint_)
+  - 🟥 [Disallow unused variables](#disallow-unused-variables-eslint) - **1 warning**
+  - 🟥 [Enforce a maximum number of lines of code in a function](#enforce-a-maximum-number-of-lines-of-code-in-a-function-eslint) - **1 warning**
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap:28]: I would prefer to keep the mock report as is, this configuration is less meaningful (e.g. why should unused variables impact performance score?). I realize the mock report doesn't cover your use case of sorting multiple groups, but I think it would be better to tests these different cases at source (i.e. in `sorting.unit.test.ts`).

File: packages/utils/perf/score-report/optimized0.ts (hunk 1 above)
Comment (Tlacenka) [reply packages/utils/perf/score-report/optimized0.ts:25]: This is in the `perf` folder. So here I had two options - either just ignore the errors or do best effort.
Since this benchmark uses the old type, I would otherwise have to refactor this function...

File: packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap (hunk 3 above)
Comment (Tlacenka) [reply packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap:28]: There is no other test that currently tests the full output of `generateMdReport` (the E2E test is very limited). So in order to check that multiple groups are correctly displayed there, at least two of them have to exist. At the moment, there is at most one which is not sufficient. I will be happy to hear better suggestions in regards to the test data.

Group sorting that involves its audits being sorted has test cases added in `utils.unit.test.ts`, namely for `getSortableGroupByRef` and `compareCategoryAuditsAndGroups`.

File: packages/utils/perf/score-report/optimized0.ts (hunk 1 above)
Comment (matejchalk) [reply packages/utils/perf/score-report/optimized0.ts:25]: Ah, I see. Then f*** it :laughing:

File: packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap (hunk 3 above)
Comment (matejchalk) [reply packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap:28]: If we were integrate this into the realistic test data, then I'd suggest the following changes:
- replace [no-unused-vars](https://eslint.org/docs/latest/rules/no-unused-vars) (unrelated to performance) with [react/jsx-key](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-key.md) (101 of React performance)
- give the ESLint group weight 0, so it doesn't impact the category score

File: testing/test-utils/src/lib/utils/dynamic-mocks/categories.mock.ts [hunk 4]
```diff
@@ -8,11 +8,17 @@ const CATEGORIES_MAP = {
...
+        plugin: 'eslint',
+        slug: 'performance',
+        weight: 0,
+      },
       {
         type: 'group',
         plugin: 'lighthouse',
         slug: 'performance',
-        weight: 1,
+        weight: 4,
```
Comment (matejchalk) [inline testing/test-utils/src/lib/utils/dynamic-mocks/categories.mock.ts:21]:
```suggestion
        weight: 1,
```

File: testing/test-utils/src/lib/utils/dynamic-mocks/eslint-plugin.mock.ts [hunk 5]
```diff
@@ -2,12 +2,28 @@ import { join } from 'node:path';
...
+  title: 'ESLint performance rules',
+  refs: [
+    {
+      slug: ESLINT_AUDITS_MAP['max-lines-per-function'].slug,
+      weight: 10,
+    },
+    {
+      slug: ESLINT_AUDITS_MAP['react-jsx-key'].slug,
+      weight: 25,
+    },
```
Comment (matejchalk) [inline testing/test-utils/src/lib/utils/dynamic-mocks/eslint-plugin.mock.ts:23]: The weights are kinda unintuitive - `max-lines-per-function` is ~28.6% and `react-jsx-key` is ~71.4%.

File: packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap [hunk 6]
```diff
@@ -24,6 +24,10 @@ Performance metrics [📖 Docs](https://developers.google.com/web/fundamentals/p
   - 🟩 [Speed Index](#speed-index-lighthouse) - **1.2 s**
 
 - 🟥 [Disallow missing \`key\` props in iterators/collection literals](#disallow-missing-key-props-in-iterators-collection-literals-eslint) (_ESLint_) - **1 warning**
+- 🔴 ESLint performance rules (_ESLint_)
+  - 🟥 [Disallow missing \`key\` props in iterators/collection literals](#disallow-missing-key-props-in-iterators-collection-literals-eslint) - **1 warning**
+  - 🟥 [Enforce a maximum number of lines of code in a function](#enforce-a-maximum-number-of-lines-of-code-in-a-function-eslint) - **1 warning**
+
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap:30]: What? 😅 where
image

File: testing/test-utils/src/lib/utils/dynamic-mocks/eslint-plugin.mock.ts (hunk 5 above)
Comment (Tlacenka) [reply testing/test-utils/src/lib/utils/dynamic-mocks/eslint-plugin.mock.ts:23]: Bruh...

File: packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap (hunk 6 above)
[the remark this answers is not in the document]
Comment (matejchalk) [reply packages/utils/src/lib/reports/__snapshots__/generate-md-report.integration.test.ts.snap:30]: Here :laughing: 

image

File: testing/test-utils/src/lib/utils/dynamic-mocks/eslint-plugin.mock.ts (hunk 5 above)
Comment (matejchalk) [reply testing/test-utils/src/lib/utils/dynamic-mocks/eslint-plugin.mock.ts]: Just give them both weight 1?

**Review by matejchalk — APPROVED** [review]
